### PR TITLE
Config networkpolicies for certmanager when using dns01

### PR DESCRIPTION
**Pull request by robinAwallace**
> [!warning]
> **This is public repository, ensure not to disclose:**
> - [x] personal data beyond what is necessary for interacting with this pull request
> - [x] business confidential information, such as customer names

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [x] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change  
- [ ] kind/dev-change    
- [ ] kind/security      
- [ ] [kind/adr]()       

### What does this PR do / why do we need this PR?

When trying out dns01 towards route53 in a full environment I noticed that cert-manager was blocked by the network policies. This allows the platform admin to configure the ip range when using dns01. 

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Network Policy checks:
  - [x] Any changed pod is covered by Network Policies
  - [x] The change does not cause any dropped packages in the `NetworkPolicy Dashboard`
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Bug checks:
  - [ ] The bug fix is covered by regression tests

### CODE REVIEW

File: config/config/common-config.yaml [hunk 1]
```diff
@@ -999,6 +1000,9 @@ networkPolicies:
     letsencrypt:
       ips:
         - set-me
+    # Enable this if dns01 is used by certManager
```
Comment (aarnq) [inline config/config/common-config.yaml:1003]: Style
```suggestion
    # Configure this if DNS-01 challenges are enabled in cert-manager
```

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -99,6 +99,8 @@ helmfile -e <service|workload>_cluster -lapp=ingress-nginx -lapp=node-local-dns
 
 > [!important]
 > To use certificates from Let's Encrypt you must enable [DNS-01 challenges in `cert-manager`](https://cert-manager.io/docs/configuration/acme/dns01/).
+> Remember to add the correct network policies.
+> For Amazon route53 the ip range is: `205.251.192.0/18`
```
Comment (aarnq) [inline DEVELOPMENT.md:103]: Style
```suggestion
> Remember to add the correct network policies, for Amazon Route 53 the IP range is: `205.251.192.0/18`.
```

Also is this range region specific, and what does it point to API or DNS resolvers?

File: helmfile.d/values/networkpolicies/common/cert-manager.yaml.gotmpl [hunk 3]
```diff
@@ -31,6 +31,15 @@ policies:
           ports:
             - tcp: 443
         {{- end }}
+        {{- with .Values | get "networkPolicies.certManager.dns01.ips" list }}
+        - name: dns01
```
Comment (aarnq) [inline helmfile.d/values/networkpolicies/common/cert-manager.yaml.gotmpl:35]: Maybe we should follow the convention (or change the convention).
```suggestion
        - name: egress-rule-dns01
```

File: helmfile.d/values/networkpolicies/common/cert-manager.yaml.gotmpl [hunk 4]
```diff
@@ -31,6 +31,15 @@ policies:
...
+          peers:
+            {{- range . }}
+            - cidr: {{ . }}
+            {{- end }}
+          ports:
+            - tcp: 53
```
Comment (aarnq) [inline helmfile.d/values/networkpolicies/common/cert-manager.yaml.gotmpl:41]: Should we include udp here as well?

**Review by aarnq — COMMENTED** [review]
Is this strictly for checking DNS propagation towards upstream or is this also intended for API access? In that case we should make that clear or open the ports for configuration so it is more flexible to use.

(My though was that cert-manager used the configured DNS resolver to check propagation but maybe they've changed that.)

File: DEVELOPMENT.md (hunk 2 above)
Comment (robinAwallace) [reply DEVELOPMENT.md:103]: Unsure, it was looking up several IP addresses. One came back as `AMAZON-05`, with no region name.

File: helmfile.d/values/networkpolicies/common/cert-manager.yaml.gotmpl (hunk 3 above)
Comment (robinAwallace) [reply helmfile.d/values/networkpolicies/common/cert-manager.yaml.gotmpl:35]: No, sounds good to change it to your suggestion.

File: helmfile.d/values/networkpolicies/common/cert-manager.yaml.gotmpl (hunk 4 above)
Comment (robinAwallace) [reply helmfile.d/values/networkpolicies/common/cert-manager.yaml.gotmpl:41]: Sure, I did not need it for route 53. But I do not see why not to add it.

**Comment by robinAwallace** [comment]
> Is this strictly for checking DNS propagation towards upstream or is this also intended for API access? In that case we should make that clear or open the ports for configuration so it is more flexible to use.
> 
> (My though was that cert-manager used the configured DNS resolver to check propagation but maybe they've changed that.)

My issue was for DNS propagation, and I followed the instructions that are for local cluster setup (as I did not find any other). So I thought it may have been missed from before. But maybe I have miss configured something? :thinking:

**Review by aarnq — APPROVED** [review]

File: config/config/common-config.yaml [hunk 5]
```diff
@@ -355,7 +355,7 @@ kubeStateMetrics:
     limits:
       cpu: 30m
       memory: 100Mi
-  # Enable this if your cluster is managed by Cluster API to get Cluster API metrics.
+  # Configure this if DNS-01 challenges are enabled in cert-manager
```
Comment (Xartos) [inline config/config/common-config.yaml:358]: Should the old comment not also be here? It's also relevant right?

Comment (robinAwallace) [reply config/config/common-config.yaml:358]: This should not have been overwritten :eyes: I will fix it, thanks for noticing :+1:

**Review by anders-elastisys — APPROVED** [review]
I was not assigned as a reviewer, but as this is something I encountered myself recently, I would like this config to be added.

**Review by Xartos — APPROVED** [review]

**Pull request merged**